Source/WebCore/platform/audio/SharedAudioDestination.cpp

```diff
@@ -0,0 +1,315 @@
+
+    void render(AudioBus* sourceBus, AudioBus* destinationBus, size_t framesToProcess, const AudioIOPosition& outputPosition) final;
+    void isPlayingDidChange() final { }
+
+    void configureRenderThread(CompletionHandler<void(bool)>&&);
+
+    unsigned m_numberOfOutputChannels;
+    float m_sampleRate;
+
+    Ref<AudioDestination> m_destination;
+    Ref<AudioBus> m_workBus;
+    Ref<WorkQueue> m_configurationQueue;
+
+    bool m_started { false };
+
+    Lock m_renderLock;
+    Semaphore m_configurationSemaphore;
+
+    // Only accessed on m_configurationQueue:
+    Vector<RefPtr<SharedAudioDestination>> m_renderers;
+
+    bool m_needsConfiguration WTF_GUARDED_BY_LOCK(m_renderLock) { true };
+    Vector<RefPtr<SharedAudioDestination>> m_newRenderers WTF_GUARDED_BY_LOCK(m_renderLock);
+
+    // Only accessed on the audio thread:
+    Vector<RefPtr<SharedAudioDestination>> m_configuredRenderers;
+};
+
+auto SharedAudioDestinationAdapter::sharedMap() -> AdapterMap&
+{
+    static MainThreadNeverDestroyed<AdapterMap> map;
+    return map;
+}
+
+Ref<SharedAudioDestinationAdapter> SharedAudioDestinationAdapter::ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction)
+{
+    auto key = std::make_tuple(numberOfOutputChannels, sampleRate);
+    auto results = sharedMap().find(key);
+    if (results != sharedMap().end()) {
+        if (auto existingAdapter = results->value.get())
+            return existingAdapter.releaseNonNull();
+    }
+
+    Ref<SharedAudioDestinationAdapter> newAdapter = adoptRef(*new SharedAudioDestinationAdapter(numberOfOutputChannels, sampleRate, WTFMove(ensureFunction)));
+    auto weakAdapter = ThreadSafeWeakPtr<SharedAudioDestinationAdapter> { newAdapter.get() };
+    sharedMap().set(key, WTFMove(weakAdapter));
+    return newAdapter;
+}
+
+SharedAudioDestinationAdapter::SharedAudioDestinationAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction)
+    : m_numberOfOutputChannels { numberOfOutputChannels }
+    , m_sampleRate { sampleRate }
+    , m_destination { ensureFunction(*this) }
+    , m_workBus { AudioBus::create(numberOfOutputChannels, AudioUtilities::renderQuantumSize).releaseNonNull() }
+    , m_configurationQueue { WorkQueue::create("SharedAudioDestinationAdapter configuration queue") }
+    , m_configurationSemaphore(0)
+{
+}
+
+SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter()
+{
+    auto key = std::make_tuple(m_numberOfOutputChannels, m_sampleRate);
+    sharedMap().remove(key);
+    m_configurationQueue->dispatchSync([] { });
+    m_destination->clearCallback();
+}
+
+void SharedAudioDestinationAdapter::addRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, weakThis = ThreadSafeWeakPtr { *this }, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
```

Comment:
Per the new smart pointer rules, we need a RefPtr *on the stack* to the object you're calling a non-trivial function on. This means, you need a local RefPtr to the WorkQueue before calling `dispatch()` on it. The way we usually do it in new code is adding a new `protectedConfigurationQueue()` member function which returns a `Ref` and call it here: `protectedConfigurationQueue()->dispatch()`.

Same comment applies below.